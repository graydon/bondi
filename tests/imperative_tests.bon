(* 
%open "../tests/imperative_tests.bon";;
*)


(* this file tests all of the operators mentioned explicitly 
	in the function eval in eval.ml
	*)

(*** eqcons *) 

Pair eqcons Pair;;
3 eqcons 3;;
'a' eqcons 'a';;

not (Pair eqcons Nil);;
not (3 eqcons 4);;
not (3 eqcons 3.0);;
not ((fun x -> x) eqcons (fun x -> x));;

(* printing tests duplicated from datum_tests.bon *) 

(* printing chars and strings  
	printchar and printstring do not quote.
	prim2string adds single or double quotes. 
	the generic print calls prim2string.
	To display text use printchar and printstring.
	To discuss text (as data) use print. 	
		*) 

printchar 'a';;
printstring (prim2string 'a');;
prim2string 'a';;
'a';;

printstring "abc";;
printstring (prim2string "abc");;
prim2string "abc";;
"abc";;

printchar '\n';;
printchar 'a';;

printstring "At present, newlines and other whitespace" ; 
printchar '\n' ; 
printstring "are only accessible as characters."  ; 
printchar '\n';
printstring "To change this, modify the lexer for strings." ;
printchar '\n'
;;



let x = Ref 3 in x = 4; x;;
let x = Ref 3 in x = 4; !x eqcons 4;;
let x = Ref 3 in x = !x+2; !x eqcons 5;;

let x = Ref 3 in 
print !x;
x = !x - 1 ; 
print !x;
x= !x - 1 ;
print !x;
printchar '\n';;

let x = Ref 5 in 
  let rec (c:Unit -> Unit) = | () -> 
    if (!x > 0) 
    then (print !x;x = !x minusint 1;c()) 
    else ()
	in c();;

let x = Ref 5 in 
let y = clone x in 
print !y;
y = !y -1;
print !y;
print !x
;;


(* cloning is shallow *) 

let x = Ref (Ref 5) in 
let y = clone x in 
print ! !y;
!y = ! !y -1;
print ! !y;
print ! !x
;;



while False do print "surprise!";;

let x = Ref 20 in 
printstring "Here are the numbers from 20 to 0:";
printchar '\n';
while !x > 0 do 
  begin 
	print !x; 
	printchar ',';
	x = !x - 1
  end ;
printchar '0';
printchar '\n'
;;


let factorial n = 
let x = Ref n in
let res = Ref 1 in 
while (!x > 1) do (
  res = !res * !x; x = !x - 1
	);
!res
;;
 
factorial 5 eqcons 120;;





(* forall is defined in "../preludes/imperatives_core.bon" *) 

for i = 0 to 3 do  print i;;

let factorial2 n = 
let res = Ref 1 in 
(for i = 1 to n do res = !res * i);
!res;;


factorial2 5 eqcons 120;;

(*** _ref *) 

_ref;;
(| _ref -> True) (Ref 3);;
(| _ref -> False | _ -> True) 3;;



(*** arrays *) 


let xs = newarray 1.0 5;;
(lengthv xs) eqcons 5;;
!(entry(xs,3)) eqcons 1.0;;
entry(xs,3) = 1.1;
!(entry(xs,3)) eqcons 1.1;;

_array;;
(| y as _array -> True) xs;;

(*** datum ops for arrays *) 

let doubleAny = 
| x as _ref  -> x = !x+ !x
| x -> ()
;;

doubleAny (entry(xs,3));
!(entry(xs,3)) eqcons 2.2;;

iter doubleAny xs;;
!(entry(xs,3)) eqcons 4.4;;

xs == xs;;

let ys = xs + xs;;
!(entry(ys,3)) eqcons 8.8;;

let ys = xs - xs;;
!(entry(ys,3)) eqcons 0.0;;

let ys = xs * xs;;
!(entry(ys,3)) eqcons (4.4 * 4.4);;

let ys = xs / xs;;
!(entry(ys,3)) eqcons 1.0;;

not (xs < xs);;
xs <= xs;;
not (xs > xs);;
xs >= xs;;


(*** generic queries for arrays *) 

apply2all (fun x -> x+x) xs;;
!(entry(xs,3));;

let (bigfloat: Float -> a -> Maybe Float) r = 
  | (x:Float) -> if x > r then (Some) x else None 
| _ -> None
;;

select (bigfloat 4.2) xs == [4.4];;
select (bigfloat 1.0) xs ==  [2.,4.4,2.,2.,2.];;



(* ref and polymorphism  *) 

let x = Ref 1 in x = 2 + !x; !x eqcons 3;;


let x = Ref [1] in x = Cons 2 !x;!x == [2,1];;




(* a type error, [] is too polymorphic to be assigned  

let x = Ref [] in 
x = Cons 1 !x;
x = Cons True !x;
!x;;

*)

let x = Ref ([]: List Int) in x = Cons 1 !x; !x == [1];;


let x = Ref 3 in 
let y = x in 
y = 4;
!x eqcons 4;;

(* but the following fails because of Top in the type of the assigned ref 

let x = 3 in 
(x:Top) =  True;
x;;

let x = Ref 3 in 
let y = (x: ref Top) in 
y = True;
x;;

*)

(* no assignment to arbitrary terms 

let fst = | (x,y) -> x;;


let x = Ref (2,3) in fst x = 4; !x;;
let x = Ref (2,3) in fst x;;
let x = Ref (2,3) in fst !x = 4; !x;;


but the following work ... 

*) 

let x = (Ref 2,Ref 3) in fst x = 4; x;;


let y = Ref 2 in 
let z = Ref 3 in 
let x = (!y,!z) in 
y = 5;
x == (2,3);;

let y = Ref 2 in 
let z = Ref 3 in 
let f  = | () -> (!y,!z) in 
y = 5;
f () == (5,3);;

