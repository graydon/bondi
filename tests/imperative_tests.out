(* Begin sourcing "./prelude/standard_prelude.bon"... *)
(* Begin sourcing "./prelude/printing.bon"... *)
^: String -> String -> String
toString: a -> String
print: a -> Unit
println: a -> Unit
(* Finished sourcing "./prelude/printing.bon" *)
(* Begin sourcing "./prelude/imperatives_core.bon"... *)
forall: Int -> Int -> (Int -> a) -> Unit
iter: (all a.a -> Unit) -> b -> Unit
clone: a -> a
(* Finished sourcing "./prelude/imperatives_core.bon" *)
(* Begin sourcing "./prelude/structure_poly.bon"... *)
Pair: a -> b -> a * b
fromTuple: a -> String
it: Unit
fst: a * b -> a
snd: a * b -> b
Inl: a -> a + b
Inr: b -> a + b
Evr: a -> Konstant a b
Ths: a -> Identity a
ParamPair: a b -> c b -> ParamProduct a c b
Ok: a b -> Okay a b
Nest: a (b c) -> Nested a b c
Tag: (a -> b) -> a c -> Represent a b c
deconstruct: a b -> (all c.Represent c a b -> d) -> d
reconstruct: Represent a b c -> b c
map: (a -> b) -> c a -> c b
Nil: List a
Cons: a -> List a -> List a
toString: b -> String
toString: c -> String
Some: a -> Maybe a
None: Maybe a
foldleft: (a -> b -> a) -> a -> c b -> a
foldright: (a -> b -> b) -> c a -> b -> b
zipwith: (a * b -> c) -> d a * e b -> d c
(* Finished sourcing "./prelude/structure_poly.bon" *)
(* Begin sourcing "./prelude/datum2generic.bon"... *)
&&: Bool -> Bool -> Bool
not: Bool -> Bool
||: Bool -> Bool -> Bool
equal: a * b -> Bool
==: a -> a -> Bool
!=: a -> a -> Bool
plus: a * b -> a
+: a -> a -> a
shortplus: a * b -> a
minus: a * b -> a
-: a -> a -> a
times: a * b -> a
*: a -> a -> a
divide: a * b -> a
/: a -> a -> a
negate: a -> a
lessthan: a * b -> Bool
<: a -> a -> Bool
lessthanorequal: a * b -> Bool
<=: a -> a -> Bool
greaterthan: a * b -> Bool
>: a -> a -> Bool
greaterthanorequal: a * b -> Bool
>=: a -> a -> Bool
(* Finished sourcing "./prelude/datum2generic.bon" *)
(* Begin sourcing "./prelude/dynamic2static.bon"... *)
dynamic_case_for_internal_use_only = ...
(* Finished sourcing "./prelude/dynamic2static.bon" *)
(* Begin sourcing "./prelude/queries.bon"... *)
apply2all: (all a.a -> a) -> b -> b
append: List a -> List a -> List a
some_append: Maybe a -> List a -> List a
select: (all a.a -> Maybe b) -> c -> List b
is: lin a -> b -> Maybe a
(* Finished sourcing "./prelude/queries.bon" *)
(* Finished sourcing "./prelude/standard_prelude.bon" *)
(* Begin sourcing "./imperative_tests.bon"... *)
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Unit
ait: Unit
'a'it: String
it = "'a'"
it: Char
it = 'a'
it: Unit
abcit: Unit
"abc"it: String
it = ""abc""
it: String
it = "abc"
it: Unit

it: Unit
ait: Unit
At present, newlines and other whitespace
are only accessible as characters.
To change this, modify the lexer for strings.
it: ref Int
it: Bool
it = True
it: Bool
it = True
it: Unit
321
it: Unit
54321it: Unit
545it: Unit
544it: Unit
it: Unit
Here are the numbers from 20 to 0:
20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1,0
factorial: Int -> Int
it: Bool
it = True
it: Unit
0123factorial2: Int -> Int
it: Bool
it = True
it: ref a
it: Bool
it = True
it: Bool
it = True
xs: array Float
xs = ...
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: array a
it = _array
it: Bool
it = True
doubleAny: a -> Unit
it: Bool
it = True
it: Unit
it: Bool
it = True
it: Bool
it = True
ys: array Float
ys = ...
it: Bool
it = True
ys: array Float
ys = ...
it: Bool
it = True
ys: array Float
ys = ...
it: Bool
it = True
ys: array Float
ys = ...
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: array Float
it = ...
it: Float
it = 4.4
bigfloat: Float -> a -> Maybe Float
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: Bool
it = True
it: ref Int * ref Int
it = (...,...)
it: Bool
it = True
it: Bool
it = True
(* Finished sourcing "./imperative_tests.bon" *)